### PR TITLE
fix: LEAP-193: Double check for upload url ip

### PR DESCRIPTION
**Pull request by triklozoid**
### PR fulfills these requirements
- [ ] Commit message(s) and PR title follows the format `[fix|feat|ci|chore|doc]: TICKET-ID: Short description of change made` ex. `fix: DEV-XXXX: Removed inconsistent code usage causing intermittent errors`
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)
- [ ] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [ ] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
_(check all that apply)_
- [ ] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [ ] Frontend

### Describe the reason for change
_(link to issue, supportive screenshots etc.)_

#### What does this fix?
_(if this is a bug fix)_

#### What is the new behavior?
_(if this is a breaking or feature change)_

#### What is the current behavior?
_(if this is a breaking or feature change)_

#### What libraries were added/updated?
_(list all with version changes)_

#### Does this change affect performance?
_(if so describe the impacts positive or negative)_

#### Does this change affect security?
_(if so describe the impacts positive or negative)_

#### What alternative approaches were there?
_(briefly list any if applicable)_

#### What feature flags were used to cover this change?
_(briefly list any if applicable)_

### Does this PR introduce a breaking change?
_(check only one)_
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [ ] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
_(check all that apply)_
- [ ] e2e
- [ ] integration
- [ ] unit

### Which logical domain(s) does this change affect?
_(for bug fixes/features, be as precise as possible. ex. Authentication, Annotation History, Review Stream etc.)_

### CODE REVIEW

File: label_studio/core/utils/io.py [hunk 1]
```diff
@@ -190,17 +190,26 @@ def validate_upload_url(url, block_local_urls=True):
 
         raise LabelStudioAPIException(f"Can't resolve hostname {domain}")
 
-    if not block_local_urls:
-        return
+    if block_local_urls:
+        validate_ip(ip)
+
+
+def validate_ip(ip):
```
Comment (jombooth) [inline label_studio/core/utils/io.py]: Type should be `validate_ip(ip: str) -> None`

File: label_studio/data_import/uploader.py [hunk 2]
```diff
@@ -133,7 +133,13 @@ def tasks_from_url(file_upload_ids, project, user, url, could_be_tasks_list):
         validate_upload_url(url, block_local_urls=settings.SSRF_PROTECTION_ENABLED)
         # Reason for #nosec: url has been validated as SSRF safe by the
         # validation check above.
-        response = requests.get(url, verify=False, headers={'Accept-Encoding': None})  # nosec
+        response = requests.get(url, verify=False, stream=True, headers={'Accept-Encoding': None})  # nosec
+
+        # second check for SSRF for prevent redirect and dns rebinding attacks
+        if settings.SSRF_PROTECTION_ENABLED:
+            response_ip = response.raw._connection.sock.getpeername()[0]
+            validate_ip(response_ip)
```
Comment (jombooth) [inline label_studio/data_import/uploader.py]: This looks a bit too low-level of an implementation to expose in app code; I think it should live in a library function that wraps requests.get and performs validation first on the uri, then makes the request, then revalidates the ip on the response, before returning the response object.

I googled around for tools we could use to help us get SSRF safety right and came across this library - it's a real shame that it's archived: https://github.com/JordanMilne/Advocate.

Lastly, did you audit for any other usages of `validate_upload_url` that should be checked for rebinding?

**Review by jombooth — CHANGES_REQUESTED** [review]
Approach LGTM but I'd like to see the details abstracted away.

File: label_studio/core/utils/io.py (hunk 1 above)
Comment (triklozoid) [reply label_studio/core/utils/io.py]: Fixed.

File: label_studio/data_import/uploader.py (hunk 2 above)
Comment (triklozoid) [reply label_studio/data_import/uploader.py]: A separate ticket created for other usages of `validate_upload_url`: LEAP-233

**Review by jombooth — APPROVED** [review]
:clap:

File: label_studio/core/utils/io.py [hunk 3]
```diff
@@ -213,3 +215,15 @@ def validate_ip(ip: str) -> None:
     for subnet in local_subnets:
         if ipaddress.ip_address(ip) in ipaddress.ip_network(subnet):
             raise InvalidUploadUrlError
+
+
+def ssrf_safe_get(*args, **kwargs):
+    # Reason for #nosec: url has been validated as SSRF safe by the
+    # validation check above.
```
Comment (jombooth) [inline label_studio/core/utils/io.py:223]: nit: this comment is a little inaccurate now, probably worth clarifying. Could we call validate_upload_url in this function too?

Comment (triklozoid) [reply label_studio/core/utils/io.py:223]: Fixed.

**Review by MihajloHoma — APPROVED** [review]

**Pull request merged**